src/serialport_win.cpp

```diff
@@ -311,6 +296,107 @@ void EIO_Write(uv_work_t* req) {
   } while (data->bufferLength > data->offset);
 }
 
+void EIO_Read(uv_work_t* req) {
+  ReadBaton* data = static_cast<ReadBaton*>(req->data);
+  data->bytesRead = 0;
+  int errorCode = ERROR_SUCCESS;
+  if (data->bufferData == NULL) {
+    data->bufferData = new char[data->bufferLength];
+	memset(data->bufferData, 0, data->bufferLength);
+  }
+
+  char* offsetPtr = data->bufferData;
+  offsetPtr += data->offset;
+
+  // Event used by GetOverlappedResult(..., TRUE) to wait for incoming data or timeout
+  // Event MUST be used if program has several simultaneous asynchronous operations
+  // on the same handle (i.e. ReadFile and WriteFile)
+  HANDLE hEvent = CreateEvent(NULL, TRUE, FALSE, NULL);
+
+  while (true) {
+    OVERLAPPED ov = {0};
+    ov.hEvent = hEvent;
+
+    // Start read operation - synchrounous or asynchronous
+    DWORD bytesReadSync = 0;
+    if (!ReadFile((HANDLE)data->fd, offsetPtr, data->bytesToRead, &bytesReadSync, &ov)) {
+      errorCode = GetLastError();
+      if (errorCode != ERROR_IO_PENDING) {
+        // Read operation error
+        if (errorCode == ERROR_OPERATION_ABORTED) {
+        } else {
+          ErrorCodeToString("Reading from COM port (ReadFile)", errorCode, data->errorString);
+          CloseHandle(hEvent);
+          return;
+        }
+        break;
+      }
+
+      // Read operation is asynchronous and is pending
+      // We MUST wait for operation completion before deallocation of OVERLAPPED struct
+      // or read data buffer
+
+      // Wait for async read operation completion or timeout
+      DWORD bytesReadAsync = 0;
+      if (!GetOverlappedResult((HANDLE)data->fd, &ov, &bytesReadAsync, TRUE)) {
+        // Read operation error
+        errorCode = GetLastError();
+        if (errorCode == ERROR_OPERATION_ABORTED) {
+        } else {
+          ErrorCodeToString("Reading from COM port (GetOverlappedResult)", errorCode, data->errorString);
+          CloseHandle(hEvent);
+          return;
+        }
+        break;
+      } else {
+        // Read operation completed asynchronously
+        data->bytesRead = bytesReadAsync;
+      }
+    } else {
+      // Read operation completed synchronously
+      data->bytesRead = bytesReadSync;
+    }
+
+    // Return data received if any
+    if (data->bytesRead > 0) {
+      break;
+    }
+  }
+
+  CloseHandle(hEvent);
+}
+
+void EIO_AfterRead(uv_work_t* req) {
+  Nan::HandleScope scope;
+  bool skipCleanup = false;
+  ReadBaton* data = static_cast<ReadBaton*>(req->data);
+
+  v8::Local<v8::Value> argv[3];
+	
+  if (data->errorString[0]) {
+    argv[0] = Nan::Error(data->errorString);
+	argv[1] = argv[2] = Nan::Undefined();
+	Sleep(100); // prevent the errors from occurring too fast
```

Comment:
Yes. Is the sleep still needed? I didn't want to take it out in case it was there for a good reason.